wireless_modules/middle_module/MQTT_Client_class.py

```diff
@@ -0,0 +1,102 @@
+# Author: Kunj (most is also borrowed from mqtt_test written by Blake)
+# Last modified: 3/08/2020
+# Note: Can remove all the Print statements once all testing is successful
+
+# Use micro pip to install packages
+import utime
+import upip
+upip.install("micropython-umqtt.simple2")
+try:
+    from umqtt.simple import MQTTClient
+except FileNotFoundError:
+    print("Error in importing library")
+
+
+class Client:
+    def __init__(self, client_id, mqtt_broker):
+        """
+        Initialises the MQTT Client
+        :param client_id: The unique client id used to initiate an MQTTClient class
+        :param mqtt_broker: A string holding domain name or IP address of the broker to connect to, to send and receive
+                            data.
+        """
+        self.client = MQTTClient(client_id, mqtt_broker)
+        self.mqtt_broker = mqtt_broker
+
+    def connect_and_subscribe(self, topics_to_subscribe, callback_func):
+        """
+        Connects to the MQTT broker and subscribes to topic in 'topics_to_subscribe'
+        :param topics_to_subscribe: An array of topics to subscribe to.
+                                    Eah element must be a string or byte literal (the latter is preferred)
+        :param callback_func: The function to be called whenever a message from the subscribed topics is received.
+        :return: True if a connection to the Broker is successfully established, otherwise False
+        """
+
+        # Provide the function to run whenever a message is received from the subscribed topics
+        self.client.set_callback(callback_func)
+        try:
+            # Connect to MQTT broker
+            self.client.connect()
+            print('Connected to %s MQTT broker' % (self.mqtt_broker))
+        except OSError:
+            return False
+
+        # Subscribe to each topic
+        for topic in topics_to_subscribe:
+            self.client.subscribe(topic)
+            print('Subscribed to %s topic' % (topic))
+
+        return True
+
+    def _to_bytes_literal(self, data):
+        """
+        Converts 'data' into a form MQTT can read
+        :param data: A string of data to convert to bytes literal
+        :return: The bytes literal version of the 'data'
+        """
+        str_data = str(data)
+        return str.encode(str_data)
+
+    def mqtt_pub(self, data, topic):
+        """
+        This functions takes care of all of the formatting and publishes 'data' on the given 'topic'. It also checks for
+        any pending message, but this should not be relied upon solely - use the check_for_message() method externally
+        when an incoming message is to be checked for/read.
+        :param data: A string of data to be sent
+        :param topic: A string representing the topic to send 'data' to.
+        :return: Nothing
+        """
+        try:
+            # convert the data to bytes literal so that MQTT can read it
```

Comment:
You've over-commented this a bit. Comments should describe _why_ you're doing something rather than _what_ your doing. For small-to-medium sized functions the overall function comment is usually sufficient.

Reply:
You mean no need for comments within the function if I have the overall function comment?

Reply:
Right. You only need to comment more on particularly complex blocks where you feel the need to explain your line of thought and why you did something that way.

Reply:
Just adding on, the code that was written in this function was already self explanatory. The docstring for this function also explained what happens in this function fairly well too. No need to comment for each line! (despite what uni tells you)